Indicate current step

alex-mo opened the pull request.

alex-mo committed:
Indicate current step via JS

katzmo commented on campaignion_foundation.js:
```diff
@@ -156,4 +156,35 @@ Drupal.behaviors.payment_slide.attach = function (context, settings) {
+Drupal.behaviors.form_step_data = {};
+Drupal.behaviors.form_step_data.attach = function (context, settings) {
+  // current step is always a number > 0 if the settings information is correct
+  let currentStep = settings.campaignion_foundation.webform.current_step || 1;
```
Although browser support is okish by now, it’s quite the paradigm shift to start using `let` and `const` in Drupal behaviors? I’d rather opt for broader support in this case and stick to `var`…

katzmo commented on campaignion_foundation.js:
```diff
@@ -156,4 +156,35 @@ Drupal.behaviors.payment_slide.attach = function (context, settings) {
+
+  if (currentStep !== previousStep) {
+    const event = new CustomEvent("changeFormStep", {
+      detail: [currentStep, previousStep],
```
I’d prefer a data object like `{current: currentStep, previous: previousStep}` as this would be unambiguous (you don’t have to remember an arbitrary order of unnamed array items).

alex-mo replied:
changed

katzmo replied:
Thanks! Is there a particular reason to keep them nested in 'details'?

katzmo replied:
Thanks! Is there a particular reason to keep them nested in 'detail'?

alex-mo replied:
Yes. jQuery.

alex-mo replied:
yes, it's the designated place for additional data in CustomEvent → https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/detail

katzmo replied:
I see, that’s a good reason :wink: (wasn’t aware you have to add that manually and it's not just the name for the object)

katzmo commented on template.php:
```diff
@@ -425,3 +425,18 @@ function _campaignion_foundation_exclude_block_from_contextual_links($module) {
+  $settings['campaignion_foundation']['webform'] = [
+    'total_steps' => $form_state['webform']['page_count'],
+    'current_step' => $form_state['webform']['page_num'],
+    'last_completed_step' => isset($form_state['webform']['page_visited']) ? $form_state['webform']['page_visited'] : 0,
```
We’re already using `??` in this theme, so I guess it would be ok to use it here too…?

alex-mo committed:
Trigger an event too when staying on the same AJAX step

alex-mo committed:
Refactor form steps event detail data

katzmo commented on template.php:
```diff
@@ -425,3 +425,18 @@ function _campaignion_foundation_exclude_block_from_contextual_links($module) {
+ * Provide the form step information in Drupal.settings for the behavior
+ * JavaScripts to pick up.
+ */
+function campaignion_foundation_form_webform_client_form_alter(&$form, &$form_state, $form_id) {
```
Could you please move the function after `campaignion_foundation_form_alter` to group all the form alters together?

katzmo replied:
A bit more up!

alex-mo replied:
Should now be correct.

katzmo replied:
I’d still move it farther up:

hook_form_alter()
hook_form_FORM_ID_alter()
hook_payment_forms_payment_form_alter()
hook_webform_component_render_alter()

Makes more sense to me than

hook_form_alter()
hook_payment_forms_payment_form_alter()
hook_webform_component_render_alter()
hook_form_FORM_ID_alter()

alex-mo replied:
Changed.

katzmo commented on campaignion_foundation.js:
```diff
@@ -156,4 +156,41 @@ Drupal.behaviors.payment_slide.attach = function (context, settings) {
+Drupal.behaviors.form_step_data = {};
+Drupal.behaviors.form_step_data.attach = function (context, settings) {
+  // current step is always a number > 0 if the settings information is correct
+  var currentStep = settings.campaignion_foundation.webform.current_step || 1;
```
This throws an error on pages without a webform.

alex-mo replied:
added try/catch for safety

katzmo replied:
To me not doing any form step stuff when there is no form would make more sense?

alex-mo replied:
Returning now from behaviour, if we need to `catch` (which indicates that there is no webform on the page)

katzmo replied:
Works now!

I’d find

```js
if (!settings.campaignion_foundation.webform) {
  return
}
var currentStep = settings.campaignion_foundation.webform.current_step || 1;
```
a bit more obvious (and it’s fewer lines of code). Are there other errors to be expected?

alex-mo replied:
`settings.campaignion_foundation` is currently only guaranteed to be set on webforms (because of `campaignion_foundation_form_webform_client_form_alter()`). so trying to access something below on pages without a webform will blow up, therefore the `try/catch` :smile:

katzmo replied:
So then

```js
if (!(settings.campaignion_foundation && settings.campaignion_foundation.webform)) {
  return
}
var currentStep = settings.campaignion_foundation.webform.current_step || 1;
```

would still be more clear, but I get why you prefer the try-catch solution…

alex-mo replied:
I think both are valid solution, I do not prefer one over the other, both are pretty clear to me. Will take your suggestion as you obviously care more about it...

alex-mo committed:
Return when no form on page

katzmo commented on campaignion_foundation.js:
```diff
@@ -156,4 +156,49 @@ Drupal.behaviors.payment_slide.attach = function (context, settings) {
+  }
+
+  currentStep = parseInt(currentStep, 10);
+  var previousStep = $("#page").attr("data-form-step") || 0;
```
Just realized: doesn’t this lead to "changeFormStep" always being triggered on the first page load? Is that expected behavior?

alex-mo replied:
Yes, expected. `if (previous < 1) {}` then we are on an initial page load. In the depending themes' handlers I check for that.

katzmo replied:
Ok. But why not check for that here instead?

alex-mo replied:
I decided to include the indication of the initial state in this event as well. Otherwise I would need/want to trigger a different event on initialization. I did not see any specific benefit for this.

katzmo replied:
I think neither "changeFormStep" nor "sameFormStep" should be triggered on initial page load because both don’t apply. Why would you need a custom event on initialization, you could use the native js load events?

alex-mo replied:
- Drupal behaviors trigger several times (please don't ask me why), but only 1 has current != previous in this logic
- it does not hurt to get the information: you can do all handler logic related to any steps changes in 1 handler and need not use a native JS load event, it's in 1 place. You get all the information currently needed in a single event.

alex-mo replied:
Please post a suggestion (event names :grin:) how to change the behavior. I still cannot see a compelling reason why to change the current state.

katzmo replied:
Not sure what you are asking me here? :thinking: 

1. I'd expect "changeFormStep" to fire when the form step changes. On initial page load, the form step does not change.
2. I fail to see *why* you’d need a form step related event on initial page load, but if you think it *is* needed, I'd prefer an additional event, e.g. "initFormStep".
3. I guess the most proper solution would be to only do anything when `context` is the webform → should only fire events when the webform changes and only once per change. (+ the init event as discussed in 2. when `context == '#document'`)

alex-mo replied:
1. I would argue it changes from nothing to something
2. agree
3. then we would still need to add code that ensures events get only fired once. that would complicate stuff. which element would be the best to e.g. call `$.once()` on?

katzmo replied:
:heart_eyes: 

Thank you! Now getting to a different form step as well as staying on the same step (e.g. due to a server side validation error) will trigger a "changeFormStep" event but I think this is fine. "Change" can also mean the form has changed (it was replaced via AJAX) but not necessarily the step we’re on.

alex-mo replied:
ah, yes. we would at least have the current/previous numbers to check for that then... :grin:

alex-mo committed:
Rework form steps events

torotil commented on template.php:
```diff
@@ -365,6 +365,21 @@ function campaignion_foundation_form_alter(&$form, $form_state, $form_id) {
+    'current_step' => $form_state['webform']['page_num'],
+    'last_completed_step' => $form_state['webform']['page_visited'] ?? 0,
+  ];
+  $form['#attached']['js'][] = ['data' => $settings, 'type' => 'setting'];
```
This will break when there are 2 webforms on the same page.

torotil reviewed (changes requested):
Having multiple webforms on one page is not a common use-case for us. That’s exactly why we’ll have forgotten about the PR, when it happens.

torotil commented on campaignion_foundation.js:
```diff
@@ -156,4 +156,67 @@ Drupal.behaviors.payment_slide.attach = function (context, settings) {
+        }
+      });
+      document.dispatchEvent(event);
+    });
```
Which events are going to be fired, when AJAX is not enabled?

alex-mo replied:
`initialFormStep` only. we need `changeFormStep` only for AJAX forms.

torotil commented:
If you ask me I would base this on the `webform-ajax-wrapper-*` and use `form [name="details[page_num]"]` to get the current step. The script could store this in a variable based on the wrapper ID. If the event is triggered on the form or webform ajax wrapper this could be used to differentiate easily between forms (eg. just listen to the right region).

In general: This needs to work without webform AJAX as well and for multiple webforms on one page. It shouldn’t break on non-standard configurations.

alex-mo committed:
Deal with multiple webforms on a single page

Only the form attached to the currently viewed node is the "main form"
with determines the current step.

torotil commented on template.php:
```diff
@@ -372,12 +372,22 @@ function campaignion_foundation_form_alter(&$form, $form_state, $form_id) {
-  $form['#attached']['js'][] = ['data' => $settings, 'type' => 'setting'];
+  // Check if the current node id is the node id the webform was attached to.
+  // This is needed to deal with the situation when there are multiple webforms
+  // on a single node.
```
For me the second paragraph would be enough explanation. I don’t think there is a need to explain what the if-condition does.

torotil commented on template.php:
```diff
@@ -372,12 +372,22 @@ function campaignion_foundation_form_alter(&$form, $form_state, $form_id) {
+  // the "main form" of the page. Only the main form determines the current
+  // step.
+  if ($node = menu_get_object()) {
+    if ($node->nid == $form['#node']->nid) {
```
These two lines could be combined:

```php
if (($node = menu_get_object()) && $node->nid == $form['#node']->nid) {
}
```
… or if you want to do something more for readability:

```php
$is_page_node = ($node = menu_get_object()) && $node->nid == $form['#node']->nid;
ìf ($is_page_node) {
}
```

torotil commented:
Since we don’t trigger events for more than one form on any page, it would be OK to keep triggering on `document`. So we could skip the second commit.

alex-mo commented:
> Since we don’t trigger events for more than one form on any page, it would be OK to keep triggering on `document`. So we could skip the second commit.

Yes, sounds reasonable. Will change.
